### PR TITLE
Add Fantom support

### PR DESCRIPTION
**Pull request by trangtran-enya**
**This branch supports both Fantom and Moonbeam L1s**

- The deployment files for Fantom Testnet are added
- The subgraphs for DTL, message relayers and SDK for Fantom Testnet are added
- The gas price oracle is simplified due to the fact that we only support BOBA as the fee token
- The SDK is fixed so that it can detect the messages from L1 to L2
- The gateway is modified so that it supports different L1 native token by passing the l1 native token symbol

### CODE REVIEW

File: ops/docker-compose-fantom.yml [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+services:
+  builder:
+    image: bobanetwork/builder:latest
+    build:
+      context: ..
+      dockerfile: ./ops/docker/Dockerfile.monorepo
+
+  l1_chain:
+    image: trangenya/lachesis:latest
```
Comment (trangtran-enya) [inline ops/docker-compose-fantom.yml]: Need a modified version of lachesis to run testnet

File: packages/message-relayer/src/service.ts [hunk 2]
```diff
@@ -540,3 +551,63 @@ export class MessageRelayerService extends BaseService<MessageRelayerOptions> {
...
+  address: string,
+  provider: ethers.providers.JsonRpcProvider
+): ethers.Contract => {
+  return new ethers.Contract(
+    address,
+    getContractInterface(name) as any,
+    provider
+  )
+}
+
```
Comment (trangtran-enya) [inline packages/message-relayer/src/service.ts]: I am unable to import this `loadContract` from `@eth-optimism/contracts`. Looks like this Relayer service using a different version of `@eth-optimism/contracts`. I don't know whether we can update this version.

File: packages/contracts/deploy/013-fund-accounts.ts [hunk 3]
```diff
@@ -55,14 +56,51 @@ const MoonbeamDev = {
...
+  chainID: 4003,
+  accounts: [
+    {
+      privateKey:
+        '0x8ffbfce79f185f4b4fb06cb3b51fcdf0fb7403b6e0d4011893ebabfc68d5187c',
+    },
+    // TODO: failed when init fund for the second account. Still need debug
+    // {
+    //   privateKey:
+    //     '0xa95aa44137b087b519cb7d2be9340f4c36709cb0d3de94ba6ba2095563a44696',
```
Comment (trangtran-enya) [inline packages/contracts/deploy/013-fund-accounts.ts]: Only the first account get a successful L1 -> L2 transfer. Others get CALL_EXCEPTION.

Comment (trangtran-enya) [reply packages/contracts/deploy/013-fund-accounts.ts]: Boba Fee Payment suite is failing due to this error.

File: packages/data-transport-layer/src/services/l1-ingestion/service.ts [hunk 4]
```diff
@@ -406,14 +407,18 @@ export class L1IngestionService extends BaseService<L1IngestionServiceOptions> {
...
       // Find all relevant events within the range.
-      const events: TypedEvent[] = await this.state.contracts[contractName]
-        .attach(eventRange.address)
-        .queryFilter(
-          this.state.contracts[contractName].filters[eventName](),
-          eventRange.fromBlock,
-          eventRange.toBlock
-        )
-
+      // The events returned by Fantom is not in order, so we need to sort them.
```
Comment (boyuan-chen) [inline packages/data-transport-layer/src/services/l1-ingestion/service.ts:410]: The events are not in order for FTM L1, so we have to re-order it.

File: packages/contracts/deploy/013-fund-accounts.ts [hunk 5]
```diff
@@ -118,8 +154,9 @@ const deployFn: DeployFunction = async (hre) => {
         L2BobaAddress,
         wallet.address,
         depositBobaAmount,
-        8_000_000,
-        '0x'
+        1_000_000,
+        '0x',
+        { gasLimit: 400_000 }
```
Comment (Boyuan-Chen) [inline packages/contracts/deploy/013-fund-accounts.ts:159]: The issue is like this: https://github.com/bobanetwork/product/issues/3

Comment (boyuan-chen) [inline packages/contracts/deploy/013-fund-accounts.ts]: We have to hardcode the gas limit for FTM. The gas limit from `eth_estimateGas` is not correct.

File: packages/contracts/deploy/013-fund-accounts.ts [hunk 6]
```diff
@@ -55,14 +55,51 @@ const MoonbeamDev = {
...
 // details of how to make that work cleanly. This is fine in the meantime.
 const deployFn: DeployFunction = async (hre) => {
   // Only execute this step if we're on the hardhat chain ID.
   const { chainId } = await hre.ethers.provider.getNetwork()
-  // if (chainId === defaultHardhatNetworkParams.chainId) {
-  if (chainId === MoonbeamDev.chainID) {
+  const testNet = supportedTestNetwork[chainId]
+  if (testNet) {
+    const gasLimitOption = testNet.chainID === 4003 ? { gasLimit: 400_000 } : {}
+    const depositL2Gas = testNet.chainID === 4003 ? 1_000_000 : 8_000_000
```
Comment (trangtran-enya) [inline packages/contracts/deploy/013-fund-accounts.ts]: Shall we put this into FantomDev and MoonbeamDev obj, so we don't need other if/else in the future, if we support other chains.

Comment (boyuan-chen) [reply packages/contracts/deploy/013-fund-accounts.ts]: Sounds good!

**Pull request merged**